File: posts/helper-prices-yfh.R
if (!exists("width", inherits = TRUE)) {
  width <- 252
}

if (!(exists("scale", inherits = TRUE) && is.list(get("scale", inherits = TRUE)))) {
  scale <- list("periods" = 252, "overlap" = 5)
}

# if (!exists("weights", inherits = TRUE)) {
#   
#   # weights <- 0.9 ^ ((width - 1):0)
#   weights <- rep(1, width)
#   
# }

status_f <- exists("factors", inherits = TRUE) && length(factors) > 0
status_t <- exists("tickers", inherits = TRUE) && length(tickers) > 0

if (status_t) {
  tickers <- unique(tickers)
} else {
  tickers <- character(0)
}

if (status_t) {
  
  data_ls <- yfhist::get_data(tickers)
  prices_xts <- xts::as.xts(yfhist::get_col(data_ls, "adjclose"))
  tickers <- colnames(prices_xts)
  
  if (status_f) {
    if (exists("returns_xts", inherits = TRUE) && (ncol(returns_xts) > 0)) {
      
      returns_xts <- do.call(merge, c(
        list(returns_xts, diff(log(prices_xts))),
        list(all = TRUE)
      ))
      colnames(returns_xts) <- c(factors, tickers)
      
    }
    
    if (exists("overlap_xts", inherits = TRUE) && (ncol(overlap_xts) > 0)) {
      
      overlap_xts <- do.call(merge, c(
        list(overlap_xts,
             roll::roll_mean(returns_xts[ , tickers], scale[["overlap"]],
                             min_obs = 1, na_restore = TRUE)),
        list(all = TRUE)
      ))
      colnames(overlap_xts) <- c(factors, tickers)
      
    } else {
      overlap_xts <- roll::roll_mean(returns_xts[ , tickers], scale[["overlap"]],
                                     min_obs = 1, na_restore = TRUE)
    }
    
  } else {
    
    returns_xts <- diff(log(prices_xts))
    overlap_xts <- roll::roll_mean(returns_xts[ , tickers], scale[["overlap"]],
                                   min_obs = 1, na_restore = TRUE)
    
  }
  
  # if (exists("factors", inherits = TRUE) && (length(factors) > 0) &&
  #     exists("overlap_xts", inherits = TRUE) && (ncol(overlap_xts) > 0)) {
  #   
  #   overlap_xts <- na.omit(overlap_xts)
  #   overlap_x_xts <- tail(overlap_xts[ , factors], width) # same dimension as `weights`
  #   overlap_y_xts <- tail(overlap_xts[ , tickers], width)
  #   
  # }
  
}

File: posts/helper-prices-yfh.py
# !pip install git+https://github.com/jasonjfoster/hist.git@main#subdirectory=python
import pandas as pd
import numpy as np
import yfhist as yfh

def exists(name):
  
  result = (name in locals()) or (name in globals())
  
  return result

def get_var(name):
  
  if name in locals():
    result = locals()[name]
  elif name in globals():
    result = globals()[name]
  else:
    raise NameError(name)
  
  return result

if not exists("width"):
  width = 252
  
if not (exists("scale") and isinstance(get_var("scale"), dict)):
  scale = {"periods": 252, "overlap": 5}

# if not exists("weights"):
#   
#   # weights = (0.9 ** np.arange(width - 1, -1, -1)).reshape((width, 1))
#   weights = np.ones((width, 1))

factors = list(get_var("factors")) if exists("factors") else []
tickers = list(get_var("tickers")) if exists("tickers") else []

status_f = len(factors) > 0
status_t = len(tickers) > 0

if (status_t):

  prices_ls = yfh.get_data(tickers)
  prices_df = yfh.get_col(prices_ls, "adjclose").set_index("index")
  tickers = prices_df.columns
  
  if (status_f):
    if (exists("returns_df") and isinstance(get_var("returns_df"), pd.DataFrame) and \
      (get_var("returns_df").shape[1] > 0)):
      
      returns_df = pd.concat([get_var("returns_df"), np.log(prices_df).diff()], axis = 1)
      returns_df.columns = list(factors) + list(tickers)
      
    else:
      returns_df = np.log(prices_df).diff()
    
    if (exists("overlap_df") and isinstance(get_var("overlap_df"), pd.DataFrame) and \
      (get_var("overlap_df").shape[1] > 0)):
        
      overlap_df = pd.concat([get_var("overlap_df"), returns_df[tickers].rolling(scale["overlap"], min_periods = 1).mean()], axis = 1)
      
    else:
      overlap_df = returns_df[tickers].rolling(scale["overlap"], min_periods = 1).mean()
    
  else:
    
    returns_df = np.log(prices_df).diff()
    overlap_df = returns_df[tickers].rolling(scale["overlap"], min_periods = 1).mean()

  if status_f and exists("overlap_df") and \
    (get_var("overlap_df").shape[1] > 0):

    overlap_df = overlap_df.dropna()
    overlap_x_df = overlap_df[factors][-width:] # same dimension as `weights`
    overlap_y_df = overlap_df[tickers][-width:]
